### DISCUSSION TITLE
error: need to decrease bstep

### DISCUSSION THREAD

**Discussion by YvonneKuiper** [cvnlab/GLMsingle#81] [Q&A]
Hello,
I have been receiving the following error when running glmsingle:
Error using fracridge (line 397)
need to decrease bstep!

What is the bstep and which factors influence it?
Any help is much appreciated!

**Comment by kendrickkay** [comment]
Hi - This error is unexpected in the sense that unless something is deeply wrong, it shouldn't be happening.  I wonder if there is any way you can help me reproduce your error so that I can understand what is happening.  Perhaps you can issue "dbstop if error" and see what the workspace is like at the point of the crash?

**Reply by YvonneKuiper** [reply]
Hi Kendrick, 
This is the workspace when te error occurs:
![Workspace_fracridge](https://user-images.githubusercontent.com/52511912/194948599-180fb7c3-bcf3-4664-93b1-0e91c656df4e.JPG)
Thanks

**Comment by kendrickkay** [comment]
Could you save the workspace to a .mat file and share it?  (e.g. save('test.mat'))   Assuming it's not too big.

**Comment by YvonneKuiper** [comment]
[bstep.zip](https://github.com/cvnlab/GLMsingle/files/9749243/bstep.zip)

**Comment by kendrickkay** [comment]
Hi - I took a look.

I traced the problem down to possibly a scaling issue.  The "y" variable (which presumably are fMRI timecourses) contains a number of voxels(?) that have very different dynamic ranges.  Basically, if I multiply the offending voxel with a simple number like 10, the error goes away.  Could there be something in how you prepare data to glmsingle that puts voxels possibly with very small dynamic ranges?  See attached figure.  We may have to trace this all the way back to your initial call to GLMestimatesingletrial.m.

Kendrick

<img width="1343" alt="image" src="https://user-images.githubusercontent.com/2728338/194970160-3e2adf78-5321-48d1-80b6-3eb29214b401.png">

**Reply by YvonneKuiper** [reply]
Thank you for having a look and for your help. It is much appreciated

I acquired a multi-echo dataset and used Tedana (multi-echo ICA) for denoising and optimal combination of the multiple echoes. Further, I pre-processed my data with a standard pre-processing pipeline and did not apply any smoothing. I only encounter the bstep issue in some participants but not all. Their images look alright. 
I did change the value of some voxels outside the brain from NaN to zero. Could this be a possible explanation?

**Comment by kendrickkay** [comment]
I see. This is helpful information.  My best guess is that some voxels in your data have very small dynamic ranges.  

Your options are (1) just comment out the "assert" line (397). The code should run fine after that. Looking at your data, I can confirm that this is likely not really going to cause any problems. (2) We have to look at your raw data to really track down the source of small dynamic ranges.

**Comment by YvonneKuiper** [comment]
Hi Kendrick,
Thanks again for your help! I traced the issue back to realignment issues between different echoes. Once I solved this, GLMsingle ran without any error messages.

**Comment by kendrickkay** [comment]
Wow, cool.  Nice on being able to identify the source of the problem.  Do you have any pics to share?  Always fun to see.
